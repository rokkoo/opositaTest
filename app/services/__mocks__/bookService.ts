jest.mock('../booksService', () => {
  return {
    BooksService: {
      getBooks: jest.fn().mockResolvedValue([
        {
          url: 'https://www.anapioficeandfire.com/api/books/1',
          name: 'A Game of Thrones',
          isbn: '978-0553103540',
          authors: ['George R. R. Martin'],
          numberOfPages: 694,
          publisher: 'Bantam Books',
          country: 'United States',
          mediaType: 'Hardcover',
          released: '1996-08-01T00:00:00',
          characters: ['https://www.anapioficeandfire.com/api/characters/2'],
          povCharacters: [
            'https://www.anapioficeandfire.com/api/characters/148',
          ],
        },
      ]),
    },
  };
});
