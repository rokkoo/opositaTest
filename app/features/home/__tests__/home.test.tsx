import * as React from 'react';
import { screen, waitFor } from '@testing-library/react-native';
import Home from '..';
import { renderWithReactQuery } from '@app/helpers/tests/renderWithReactQuery';
import { AppProviders } from '@app/helpers/tests/appProviders';
import { mockBooksData } from '@app/services/__mocks__/bookService';

describe('Home screen tests', () => {
  test('it should render correctly', async () => {
    renderWithReactQuery(
      <AppProviders>
        <Home />
      </AppProviders>,
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('it should show screen title', async () => {
    renderWithReactQuery(
      <AppProviders>
        <Home />
      </AppProviders>,
    );

    await waitFor(async () => {
      const output = await screen.findByTestId('HomeScreenTitle');

      expect(output).toHaveTextContent('OpositaTest');
    });
  });

  test('it should be Game of Thrones on the screen', async () => {
    renderWithReactQuery(
      <AppProviders>
        <Home />
      </AppProviders>,
    );

    await waitFor(async () => {
      const text = screen.queryByText('A Game of Thrones');
      expect(text).toBeTruthy();
    });
  });

  test('favorite section should not be displayed', async () => {
    renderWithReactQuery(
      <AppProviders>
        <Home />
      </AppProviders>,
    );

    await waitFor(async () => {
      const text = screen.queryByTestId('favoritesBooksSection');
      expect(text).toBeFalsy();
    });
  });

  test('recent book section should not be displayed', async () => {
    renderWithReactQuery(
      <AppProviders>
        <Home />
      </AppProviders>,
    );

    await waitFor(async () => {
      const text = screen.queryByTestId('recentsBooksSection');
      expect(text).toBeFalsy();
    });
  });

  test('should render favorite section', async () => {
    const favoriteContextMock = {
      books: mockBooksData,
      addBook: jest.fn(),
      removeBook: jest.fn(),
      hasFavorireBooks: true,
    };

    renderWithReactQuery(
      <AppProviders favoriteContext={favoriteContextMock}>
        <Home />
      </AppProviders>,
    );

    await waitFor(async () => {
      const text = screen.queryByTestId('favoritesBooksSection');
      expect(text).toBeTruthy();

      const bookItems = screen.getAllByTestId('book-item');
      expect(bookItems).toHaveLength(mockBooksData.length);
    });
  });

  test('should render recent visited section', async () => {
    const recentVisitedContext = {
      books: mockBooksData,
      addBook: jest.fn(),
      removeBook: jest.fn(),
      hasRecentsVisitedBooks: true,
    };

    renderWithReactQuery(
      <AppProviders recentVisitedContext={recentVisitedContext}>
        <Home />
      </AppProviders>,
    );

    await waitFor(async () => {
      const text = screen.queryByTestId('recentsBooksSection');
      expect(text).toBeTruthy();

      const bookItems = screen.getAllByTestId('recent-book-item');
      expect(bookItems).toHaveLength(mockBooksData.length);
    });
  });

  test('should render all books', async () => {
    renderWithReactQuery(
      <AppProviders>
        <Home />
      </AppProviders>,
    );

    await waitFor(async () => {
      const text = screen.queryByTestId('bookListSection');
      expect(text).toBeTruthy();

      const bookItems = screen.getAllByTestId('list-book-item');
      expect(bookItems).toHaveLength(mockBooksData.length);
    });
  });
});
