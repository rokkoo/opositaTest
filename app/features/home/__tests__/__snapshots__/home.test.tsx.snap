// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen tests it should render correctly 1`] = `
<RCTScrollView
  style={
    {
      "backgroundColor": "#F7F7F7",
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        {
          "paddingBottom": 24,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 24,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "black",
              "fontFamily": "System",
            },
            {
              "color": "black",
              "fontSize": 32,
              "fontWeight": "normal",
            },
            {
              "textAlign": "center",
            },
          ]
        }
        testID="HomeScreenTitle"
      >
        OpositaTest
      </Text>
      <View
        style={
          [
            {
              "paddingTop": 24,
            },
          ]
        }
      />
      <View
        style={
          {
            "paddingVertical": 16,
          }
        }
        testID="bookListSection"
      >
        <Text
          style={
            [
              {
                "color": "black",
                "fontFamily": "System",
              },
              {
                "color": "black",
                "fontSize": 16,
                "fontWeight": "bold",
              },
              undefined,
            ]
          }
        >
          Todos los libros
        </Text>
        <View
          style={
            [
              {
                "paddingTop": 16,
              },
            ]
          }
        />
        <RCTScrollView
          data={
            [
              {
                "authors": [
                  "George R. R. Martin",
                ],
                "characters": [
                  "https://www.anapioficeandfire.com/api/characters/2",
                ],
                "country": "United States",
                "isbn": "978-0553103540",
                "mediaType": "Hardcover",
                "name": "A Game of Thrones",
                "numberOfPages": 694,
                "povCharacters": [
                  "https://www.anapioficeandfire.com/api/characters/148",
                ],
                "publisher": "Bantam Books",
                "released": "1996-08-01T00:00:00",
                "url": "https://www.anapioficeandfire.com/api/books/1",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={true}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          showsHorizontalScrollIndicator={false}
          stickyHeaderIndices={[]}
          viewabilityConfigCallbackPairs={[]}
        >
          <View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  null,
                ]
              }
            >
              <View
                style={
                  {
                    "marginLeft": 0,
                  }
                }
                testID="list-book-item"
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                >
                  <Image
                    height={200}
                    onError={[Function]}
                    source={
                      {
                        "uri": "https://covers.openlibrary.org/b/isbn/978-0553103540-M.jpg",
                      }
                    }
                    style={
                      {
                        "borderRadius": 8,
                        "objectFit": "cover",
                      }
                    }
                    testID="img978-0553103540"
                    width={250}
                  />
                  <View
                    style={
                      {
                        "backgroundColor": "rgba(0,0,0,0.5)",
                        "borderBottomLeftRadius": 8,
                        "borderBottomRightRadius": 8,
                        "bottom": 0,
                        "height": 30,
                        "justifyContent": "center",
                        "left": 0,
                        "paddingHorizontal": 8,
                        "position": "absolute",
                        "width": "100%",
                      }
                    }
                  >
                    <Text
                      numberOfLines={1}
                      style={
                        [
                          {
                            "color": "black",
                            "fontFamily": "System",
                          },
                          {
                            "color": "#F7F7F7",
                            "fontSize": 12,
                            "fontWeight": "bold",
                          },
                          undefined,
                        ]
                      }
                    >
                      A Game of Thrones
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
